nuxeo-ai-core/src/main/java/org/nuxeo/ai/configuration/ThresholdComponent.java

```diff
@@ -0,0 +1,169 @@
+
+    public static final String AUTO_CORRECT_DEFAULT_VALUE = "nuxeo.ai.autocorrect.default.threshold";
+
+    protected float globalThreshold = 0.1f;
+
+    protected float globalAutofillThreshold = 0.1f;
+
+    protected float globalAutocorrectThreshold = 0.1f;
+
+    protected final Map<String, Map<String, Threshold>> typeThresholds = new HashMap<>();
+
+    @Override
+    public void start(ComponentContext context) {
+        super.start(context);
+        setDefaultThresholds();
+
+        List<Descriptor> descriptors = getDescriptors(THRESHOLD_CONFIGURATION_XP);
+        for (Descriptor d : descriptors) {
+            ThresholdConfiguratorDescriptor descriptor = (ThresholdConfiguratorDescriptor) d;
+            float global = descriptor.getGlobal();
+            globalThreshold = Math.max(globalThreshold, global);
+
+            List<ThresholdConfiguratorDescriptor.Configuration> configurations = descriptor.getConfigurations();
+
+            for (ThresholdConfiguratorDescriptor.Configuration config : configurations) {
+                String type = config.getType();
+                typeThresholds.computeIfAbsent(type, key -> new HashMap<>());
+
+                for (Threshold t : config.getThresholds()) {
+                    Map<String, Threshold> map = typeThresholds.get(type);
+                    t.converge(map.get(t.getXPath()));
+                    map.put(t.getXPath(), t);
+                }
+            }
+        }
+    }
+
+    @Override
+    public float getAutoFillThreshold(DocumentModel doc, String xpath) {
+        float result = -1.f;
+
+        Set<String> keys = typeThresholds.keySet();
+        if (keys.contains(doc.getType())) {
+            Map<String, Threshold> typeTshlds = typeThresholds.get(doc.getType());
+            if (typeTshlds.containsKey(xpath)) {
+                result = typeTshlds.get(xpath).getAutofillValue();
+            }
+        }
+
+        Set<String> facets = doc.getFacets();
+        boolean hasFacet = !Collections.disjoint(keys, facets);
+        if (hasFacet) {
+            for (String facet : facets) {
+                if (keys.contains(facet) && typeThresholds.get(facet).containsKey(xpath)) {
+                    result = Math.max(result, typeThresholds.get(facet).get(xpath).getAutofillValue());
+                }
+            }
+        }
+
+        return result > 0.f ? result : globalThreshold;
+    }
+
+
+    @Override
+    public float getAutoCorrectThreshold(DocumentModel doc, String xpath) {
+        float result = -1.f;
+
+        Set<String> keys = typeThresholds.keySet();
+        if (keys.contains(doc.getType())) {
+            Map<String, Threshold> typeTshlds = typeThresholds.get(doc.getType());
+            if (typeTshlds.containsKey(xpath)) {
+                result = typeTshlds.get(xpath).getAutocorrect();
+            }
+        }
+
+        Set<String> facets = doc.getFacets();
```

Comment:
You suppose Facets here, how do you set a facet threshold?

Reply:
mm on line 131 I set it.